src/images.rs

```diff
@@ -0,0 +1,254 @@
+    pub image_dir: String,
+    pub image_size: u64,
+    pub chunk_size: u64,
+    pub num_chunks: u64,
+}
+
+impl ImageReader {
+    /// Create a new chunk reader for an image.
+    pub fn new(image_name: String, image_dir: String, chunk_size: u64) -> Result<Self, Error> {
+        let meta = fs::metadata(&format!("{}/{}", image_dir, image_name))?;
+        Ok(ImageReader {
+            image_name: image_name,
+            image_dir: image_dir,
+            image_size: meta.len(),
+            chunk_size: chunk_size,
+            num_chunks: (meta.len() as f64 / chunk_size as f64).ceil() as u64
+        })
+    }
+
+    /// Read a chunk of the image at the given index.
+    pub fn read_chunk(&self, index: u64) -> Result<Vec<u8>, Error> {
+        if index >= self.num_chunks {
+            return Err(Error::Image(format!("invalid chunk index: {}", index)));
+        }
+        let mut buf = Vec::with_capacity(self.chunk_size as usize);
+        let mut file = File::open(&format!("{}/{}", self.image_dir, self.image_name))?;
+        file.seek(SeekFrom::Start(index * self.chunk_size))?;
+        file.take(self.chunk_size).read_to_end(&mut buf)?;
+        if index == self.num_chunks - 1 { // last chunk
+            buf.truncate((self.image_size % self.chunk_size) as usize);
+        }
+        Ok(buf)
+    }
+
+    /// Generate a SHA1 checksum of the image data.
+    pub fn sha1sum(&self) -> Result<String, Error> {
+        let mut hash = Sha1::new();
+        for index in 0..self.num_chunks {
+            hash.input(&self.read_chunk(index)?);
+        }
+        Ok(hash.result_str())
+    }
+
+    /// Generate metadata about the image.
+    pub fn image_meta(&self) -> Result<ImageMeta, Error> {
+        Ok(ImageMeta {
+            image_name: self.image_name.clone(),
+            sha1sum: self.sha1sum()?,
+            num_chunks: self.num_chunks,
+        })
+    }
+}
+
+
+/// Re-build an image from individual chunks.
+#[derive(Serialize, Deserialize)]
+pub struct ImageWriter {
+    pub meta: ImageMeta,
+    pub images_dir: String,
+    pub chunks_dir: String,
+    pub last_written: DateTime<Utc>,
+    pub chunks_written: HashSet<u64>,
+}
+
+impl ImageWriter {
+    /// Prepare to read an image file in chunks.
+    pub fn new(meta: ImageMeta, images_dir: String) -> Self {
+        let chunks_dir = format!("{}/chunks/{}", images_dir, meta.image_name);
+        fs::create_dir_all(&chunks_dir).expect("couldn't create chunks dir");
+        ImageWriter {
+            meta: meta,
+            images_dir: images_dir,
+            chunks_dir: chunks_dir,
+            last_written: Utc::now(),
+            chunks_written: HashSet::new(),
+        }
+    }
+
+    /// Write a chunk of the image to the disk.
+    pub fn write_chunk(&mut self, data: &[u8], index: u64) -> Result<(), Error> {
+        trace!("writing chunk {} for {}", index, self.meta.image_name);
+        let mut file = File::create(format!("{}/{}", self.chunks_dir, index))?;
```

Comment:
Does this error if the file already exists? And if not, it should ensure it is truncated.